feat: 질문 세트 관련 api

12xii opened the pull request:
- 질문 세트 등록
- 질문 세트 목록 조회
- 질문 세트 상세보기
- 질문 세트 풀기
- 질문 세트 내 질문 풀기

12xii committed:
feat: 질문 세트 등록 api 추가

12xii committed:
feat: 질문 세트 목록 조회 api 추가

12xii committed:
fix: 오타 수정

12xii committed:
feat: 질문 세트 상세 조회 api 추가

12xii committed:
feat: 질문 세트 풀이 api 추가

12xii committed:
feat: 질문 세트 내 질문 풀이 api 추가

12xii committed:
fix: 오타 수정

rlaisqls commented on src/main/kotlin/kr/hs/dsm/inq/domain/question/persistence/repository/QuestionSetsRepository.kt:
```diff
@@ -1,7 +1,142 @@
+                            /* job = */ author.job,
+                            /* jobDuration = */ author.jobDuration,
+                            /* tagList = */ GroupBy.list(QTags.tags),
+                            /* isAnswered = */ questionSets.isNull(),
```
isAnswered 필드가 'user가 문제 세트를 풀었는지 여부'를 나타내지 않는 것 같습니다

12xii replied:
수정하였습니다 확인부탁드립니다!

rlaisqls commented on src/main/kotlin/kr/hs/dsm/inq/domain/question/service/QuestionService.kt:
```diff
@@ -299,4 +281,138 @@ class QuestionService(
+                    category = null,
+                )
+            )
+        }
```
category null로 넣지 말고 request.category로 동일하게 넣어주세요. (Tags 엔티티의 category 필드도 null 불가능하게 다시 바꿔주셨음 합니당)
그리고 요청된 태그가 이미 존재하는 태그인 경우에는 조회해온 엔티티를 대입해야합니다. 질문 생성 api 참고해서 수정해주세요

rlaisqls commented on src/main/kotlin/kr/hs/dsm/inq/domain/question/service/QuestionService.kt:
```diff
@@ -299,4 +281,138 @@ class QuestionService(
+
+        val questionSet = questionSetsRepository.findByIdOrNull(questionSetID)?: throw QuestionNotFoundException
+
+        postRepository.save(Post())
```
이 부분에서는 history를 저장하는게 아니라서 post를 생성할 필요가 없을 것 같아요

rlaisqls reviewed (commented):
전반적으로 `ID`로 되어있는 부분 `Id`로 통일해주세요!

rlaisqls commented on src/main/kotlin/kr/hs/dsm/inq/domain/question/service/QuestionService.kt:
```diff
@@ -299,4 +281,138 @@ class QuestionService(
+
+        val len = request.questionId.size - 1
+        val categories: MutableList<CategoriesDto> = mutableListOf()
+        for(i: Int in 0..len) {
+            val questionId = request.questionId[i]
+
+            val question = questionsRepository.findByIdOrNull(questionId) ?: throw QuestionNotFoundException
+
+            setQuestionRepository.save(
+                SetQuestion(
+                    id = SetQuestionID(
+                        setId = sets.id,
+                        questionId = questionId,
+                    ),
+                    setId = sets,
+                    questionId = question,
+                    questionIndex = i + 1,
+                )
+            )
```
이 부분 이런식으로 수정하면 어떨까요?
```kotlin
request.questionId.map { questionId -> 
   // 반복문 내용
}
```

12xii replied:
인덱스가 필요하여 mapIndexed로 수정하였으나 mapIndexed 내부의 코드가 실행되지 않음을 확인했습니다

rlaisqls replied:
헉 forEachIndexed를 사용하면 코드가 실행될 것 같아요

12xii replied:
수정했습니당

12xii committed:
fix: tags

- registerQuestionSet 태그 저장 로직 변경
- Tags Entity category field nullable:false로 변경

12xii committed:
fix: 사용하지 않는 post() 생성 제거

12xii committed:
fix: user의 문제 풀이 여부 응답 수정

12xii committed:
fix: 반복문 수정

- for -> forEachIndexed

12xii committed:
fix: 표기 통일

12xii committed:
fix: 빌드 에러 수정

rlaisqls commented on src/main/kotlin/kr/hs/dsm/inq/domain/question/service/QuestionService.kt:
```diff
@@ -353,15 +351,19 @@ class QuestionService(
         }
 
+        val count = categories.groupingBy { it.category }.eachCount()
+        val category: MutableList<CategoriesDto> = mutableListOf()
+        count.map {
+            category.add(CategoriesDto(it.key, it.value))
+        }
```
CategoriesDto 리스트를 표현하는데 categories와 category라는 변수를 각각 다른 용도로 사용하니까 흐름을 파악하기 어려운 것 같습니다.
변수 이름이 역할을 더 직관적으로 나타내도록 수정하거나, map 등 collection 함수를 적극 활용해서 가급적이면 mutableList를 사용하지 않고 로직을 짜주셨으면 합니다.

12xii replied:
MutableList를 사용하지 않고 구현할 방법을 찾지 못하여 임시적으로 MutableList를 사용하여 구현하였습니다. 변수명은 수정하였으니 확인 부탁드립니다.

rlaisqls commented on src/main/kotlin/kr/hs/dsm/inq/domain/question/service/QuestionService.kt:
```diff
@@ -299,4 +309,133 @@ class QuestionService(
+            index, questionId -> run {
+                val question = questionsRepository.findByIdOrNull(questionId) ?: throw QuestionNotFoundException
+
+                setQuestionRepository.save(
+                    SetQuestion(
+                        id = SetQuestionId(
+                            setId = sets.id,
+                            questionId = questionId,
+                        ),
+                        setId = sets,
+                        questionId = question,
+                        questionIndex = index,
+                    )
```
setId에 set 엔티티를 넣는다는게 좀 직관적이지 않은 것 같아요
필드명을 set과 question으로 수정하면 어떨까요?

rlaisqls commented on src/main/kotlin/kr/hs/dsm/inq/domain/question/service/QuestionService.kt:
```diff
@@ -299,4 +309,133 @@ class QuestionService(
+        val categories: MutableList<CategoriesDto> = mutableListOf()
+        request.questionId.forEachIndexed {
+            index, questionId -> run {
+                val question = questionsRepository.findByIdOrNull(questionId) ?: throw QuestionNotFoundException
```
이렇게 돌리면 api를 호출할 때 너무 많은 쿼리가 생길 것 같아요... request로 주어진 id에 해당하는 question을 한번에 불러와서 사용하면 어떨까요?

12xii replied:
문제상황을 이해하였으나 해당 피드백을 이해하지 못하였습니다. 추가적인 설명 부탁드릴 수 있을까요?

rlaisqls replied:
`questionsRepository.findById(questionIds)`
이런식으로 id 목록을 넣어 여러 엔티티 한 쿼리에 가져올 수 있습니당

rlaisqls commented on src/main/kotlin/kr/hs/dsm/inq/domain/question/service/QuestionService.kt:
```diff
@@ -116,8 +91,36 @@ class QuestionService(
+            }
+        )
+    }
+    private fun saveTag(
+        request: QuestionSetsRequest,
+        problems: Problem,
+    ) {
+        val existsTagList = tagsRepository.findByCategoryAndTagIn(request.category, request.tag)
+        val existsTagMap = existsTagList.associateBy { it.tag }
+        val tagListToSave = request.tag
+            .mapNotNull { if (existsTagMap[it] != null) null else it }
+
+        val tagList = tagsRepository.saveAll(
+            tagListToSave.map {
+                Tags(
+                    tag = it,
+                    category = request.category
+                )
+            }
+        ).union(existsTagList)
+
+        questionTagsRepository.saveAll(
+            tagList.map {
+                QuestionTags(
+                    id = QuestionTagsId(problemId = problems.id, tagId = it.id),
+                    problems = problems,
```
파라미터에서 QuestionSetsRequest 말고 problems와 taglist를 받게 해서, question 생성 api에 쓰이는 함수랑 합쳐주실 수 있나요?

rlaisqls reviewed (commented):
가독성이나 디테일면에서 개선할 수 있는 부분이 존재하는데, 만약 고칠 시간이 없으시다면 그냥 넘어가도.. 큰 상관은 없습니다.
각 코멘트 한번씩만 확인해주세요

12xii committed:
fix: 피드백 반영

- category & categories => 변수명 수정 및 통합
  - MutableList를 사용하지 않고 구현하는 방법을 고민해보았으나 생각하지 못하여 MutableList 사용
- saveTag 함수 수정 및 통합

12xii committed:
fix: 코드 깔끔하게 수정

12xii committed:
Merge branch 'main' into question-set
